### PR TITLE
[new feature] Simulating future prices of a stock based on historical data

### PR DESCRIPTION
**Pull request by matijhs**
### [new feature] Simulating future prices of a stock based on historical data

The goal of the implemented ChallengeService::projectAssetData is to project future stock prices by generating different _Scenarios_ for an asset based on it's historical price changes and return the best one out of them. 

- Created a  _Scenario_ object for holding together the start/end trading dates, open/close prices and the generated montly price projections.
- **ScenarioGenerator::generateScenario** - Returns a list of _Scenarios_ by applying random changes from the given price history onto the current price _(current price now is the last element of the price history, which gives wrong current price if the history is not the latest)_.
- Price history is collected by calling the _/{symbol}/historical_  endpoint _(hardcoded to collect the last 5 years of data, this could be improved)_.
- **ScenarioGenerator::extractMonthlyPriceChange**s - Extracts a list of month-over-month price changes from the available historical data using only the closePrice field of pricings. 
- **ChallengeService::getProjectedAssetData** - Has a tunable parameter _NUMBER_OF_SCENARIOS_TO_GENERATE = 1000_ which controls the amount of _Scenarios_ to generate.
- Extended the _/{asset}/projected_ endpoint by an optional _numberOfMonths_ request parameter that controls how many months of future data to generate. The default value is set to 12 months, the minimum is 1 month.

### CODE REVIEW

File: src/main/java/com/ndvr/challenge/util/Utils.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+public class Utils {
+
+    /**
+     * Returns the median closePrice value of a Scenario.
+     * @param sortedScenarios - A list of Scenarios sorted by their closePrice value in ascending order.
+     * @return Returns the middle element, if the number of elements in sortedScenarios are odd.
+     * Otherwise, it returns the average value of the elements at the 'middle' and 'middle-1' indexes.
+     */
+    public static BigDecimal Median(List<Scenario> sortedScenarios) {
```
Comment (csirmazbendeguz) [inline src/main/java/com/ndvr/challenge/util/Utils.java:17]: fuggvenyeket szokas nevezni igy Java-ban? a getMedian nevet vartam volna

File: src/test/java/com/ndvr/challenge/util/UtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+public class UtilsTest {
+    @Test
+    void testMedianForOddAndEvenNumberOfElements() {
+        List<Scenario> scenarios = new java.util.ArrayList<>(List.of(
+                new Scenario("A", BigDecimal.valueOf(10)),
+                new Scenario("B", BigDecimal.valueOf(20)),
+                new Scenario("C", BigDecimal.valueOf(30))));
+
+        assertEquals(20, Median(scenarios).intValue());
+        scenarios.add(new Scenario("D", BigDecimal.valueOf(40)));
```
Comment (csirmazbendeguz) [inline src/test/java/com/ndvr/challenge/util/UtilsTest.java:20]: elegansabb lenne erre kulon tesztesetet irni

File: src/test/java/com/ndvr/challenge/util/UtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import org.junit.jupiter.api.Test;
+import java.math.BigDecimal;
+import java.util.List;
+
+import static com.ndvr.challenge.util.Utils.Median;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class UtilsTest {
+    @Test
+    void testMedianForOddAndEvenNumberOfElements() {
```
Comment (csirmazbendeguz) [inline src/test/java/com/ndvr/challenge/util/UtilsTest.java:13]: esetleg ures listat is tesztelhetnet

File: src/main/java/com/ndvr/challenge/generator/ScenarioGenerator.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        return scenario;
+    }
+
+     static List<BigDecimal> extractMonthlyPriceChanges(List<Pricing> pricingHistory) {
+        List<BigDecimal> priceChanges = new ArrayList<>();
+        int startIndex = 0;
+
+        for (int endIndex = 0; endIndex < pricingHistory.size(); endIndex++) {
+            if (pricingHistory.get(endIndex).getTradeDate()
+                    .isAfter(pricingHistory.get(startIndex).getTradeDate().with(lastDayOfMonth()))) {
```
Comment (csirmazbendeguz) [inline src/main/java/com/ndvr/challenge/generator/ScenarioGenerator.java:51]: ezek a sorok kicsit hosszuak. jo lenne egy valtozo, e.g.
`startOfMonthPricing = pricingHistory.get(startIndex);`
`endOfMonthPricing = pricingHistory.get(endIndex);`

File: src/main/java/com/ndvr/challenge/generator/ScenarioGenerator.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    private static BigDecimal subtract(Pricing endPricing, Pricing startPricing) {
+        return endPricing.getClosePrice().subtract(startPricing.getClosePrice());
+    }
+
+    /**
+     * Returns a random element from the given priceChangeList.
+     * @param priceChangeList - A list of price changes.
+     * @return Random element of priceChangeList.
+     */
+    private static BigDecimal getRandomPriceChange(List<BigDecimal> priceChangeList){
```
Comment (csirmazbendeguz) [inline src/main/java/com/ndvr/challenge/generator/ScenarioGenerator.java:73]: lehetne generic akar

File: src/main/java/com/ndvr/challenge/generator/ScenarioGenerator.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            // Edge case if the last history element is in the same month
+            else if (endIndex == pricingHistory.size() - 1) {
+                priceChanges.add(subtract(pricingHistory.get(endIndex), pricingHistory.get(startIndex)));
+                startIndex = endIndex;
+            }
+        }
+        return priceChanges;
+    }
+
+    private static BigDecimal subtract(Pricing endPricing, Pricing startPricing) {
```
Comment (csirmazbendeguz) [inline src/main/java/com/ndvr/challenge/generator/ScenarioGenerator.java:64]: ez a fuggveny nem lenne jobb a Pricing osztalyban?

File: src/main/java/com/ndvr/challenge/generator/ScenarioGenerator.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+public class ScenarioGenerator {
+
+    /**
+     * Generates a 'Scenario' by calculating a list of monthly prices based on the currentPricing, the priceChangeList
+     * and the given numberOfMonths into the future.
+     * @param numberOfMonths - The time until the price projection will be made into the future.
+     * @return An instance of Scenario holding the startDate, endDate, startPrice, closePrice and a list of projected future monthly prices.
+     */
+    public static Scenario generateScenario(String symbol, List<Pricing> priceHistory, int numberOfMonths) {
+        Pricing currentPricing = priceHistory.get(priceHistory.size() -1);
```
Comment (csirmazbendeguz) [inline src/main/java/com/ndvr/challenge/generator/ScenarioGenerator.java:23]: space hianyzik

File: src/main/java/com/ndvr/challenge/generator/ScenarioGenerator.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        BigDecimal currentPrice = currentPricing.getClosePrice();
+        LocalDate currentDate = currentPricing.getTradeDate();
+
+        Scenario scenario = new Scenario(symbol);
+        scenario.setOpenPrice(currentPrice);
+        scenario.setStartDate(currentDate);
+        scenario.setEndDate(currentPricing.getTradeDate().plusMonths(numberOfMonths));
```
Comment (csirmazbendeguz) [inline src/main/java/com/ndvr/challenge/generator/ScenarioGenerator.java:30]: ujrahasznalhatnad a currentDate valtozot

File: src/test/java/com/ndvr/challenge/generator/ScenarioGeneratorTest.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    void testGenerateScenarioReturnsCompleteScenarioObject() {
+        List<Pricing> priceHistory = List.of(
+                price1, price2, price3, price4, price5
+        );
+        int monthToAdd = 10;
+
+        Scenario scenario = ScenarioGenerator.generateScenario("ABC", priceHistory, monthToAdd);
+
+        assertEquals(monthToAdd, scenario.getProjectedFutureMonthlyPrices().size());
+        assertNotNull(scenario.getStartDate());
```
Comment (csirmazbendeguz) [inline src/test/java/com/ndvr/challenge/generator/ScenarioGeneratorTest.java:56]: jo lenne tesztelni az ertekeket, nem csak azt, hogy nem null

File: src/test/java/com/ndvr/challenge/generator/ScenarioGeneratorTest.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+                    LocalDate.of(2012,3,5));
+
+    private final Pricing price5 =
+            new Pricing(null,
+                    BigDecimal.valueOf(15.8),
+                    null, null,
+                    LocalDate.of(2012,3,20));
+
+    @Test
+    void testGenerateScenarioReturnsCompleteScenarioObject() {
```
Comment (csirmazbendeguz) [inline src/test/java/com/ndvr/challenge/generator/ScenarioGeneratorTest.java:47]: tesztelni kellene ures listara is

File: src/main/java/com/ndvr/challenge/service/ChallengeService.java [hunk 11]
```diff
@@ -2,30 +2,49 @@
...
-    
-    public List<BigDecimal> getProjectedAssetData(String symbol) {
-        log.info("Generating projected price data for {}", symbol);
-        // TODO Implement getProjectedAssetData()
-        return List.of();
-    }
 
+    public List<BigDecimal> getProjectedAssetData(String symbol, int numberOfMonths) {
+        log.info("Generating projected price data for {} {} months into the future", symbol, numberOfMonths);
+        List<Pricing> priceHistory =
```
Comment (csirmazbendeguz) [inline src/main/java/com/ndvr/challenge/service/ChallengeService.java:33]: ez siman lehet ures, mert a kliens csak warning-ot dob es ures listat ad vissza. ezt az esetet kezelni kell

File: src/test/java/com/ndvr/challenge/generator/ScenarioGeneratorTest.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    void testExtractMonthlyPriceChangesReturnsCorrectValues() {
+        List<BigDecimal> priceChanges = extractMonthlyPriceChanges(List.of(price1, price2, price3, price4));
+
+        assertEquals( 2, priceChanges.size());
+        assertEquals(0.1, (priceChanges.get(0).doubleValue()));
+        assertEquals(0.1, (priceChanges.get(1).doubleValue()));
+    }
+
+    @Test
+    void testExtractMonthlyPriceChangesReturnsCorrectValuesForEdgeCase() {
```
Comment (csirmazbendeguz) [inline src/test/java/com/ndvr/challenge/generator/ScenarioGeneratorTest.java:72]: ez a fuggveny nev nem tul jo, az "edge case" az utalhat barmire

File: src/test/java/com/ndvr/challenge/generator/ScenarioGeneratorTest.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        List<BigDecimal> priceChanges = extractMonthlyPriceChanges(List.of(price1, price2, price3));
+
+        // last price should be ignored
+        assertEquals( 1, priceChanges.size());
+        assertEquals(0.1, (priceChanges.get(0).doubleValue()));
+    }
+
+    @Test
+    void testExtractMonthlyPriceChangesWithMoreDates() {
```
Comment (csirmazbendeguz) [inline src/test/java/com/ndvr/challenge/generator/ScenarioGeneratorTest.java:81]: mi van akkor, ha van 1+ honap kihagyas az adatban? ezt is kezelni kene + tesztelni.
garantalt, hogy a lista rendezve van? kivetelt lehetne dobni ha nem.

File: src/main/java/com/ndvr/challenge/service/ChallengeService.java [hunk 14]
```diff
@@ -2,30 +2,49 @@
...
+                this.getHistoricalAssetData(symbol, LocalDate.now().minusYears(5), LocalDate.now());
```
Comment (csirmazbendeguz) [inline src/main/java/com/ndvr/challenge/service/ChallengeService.java:34]: kicsit random ez a 5 itt, ez miert 5 ev?

File: src/main/java/com/ndvr/challenge/service/ChallengeService.java [hunk 15]
```diff
@@ -2,30 +2,49 @@
...
         return dataProvider.fetchPriceData(symbol, fromDate, toDate);
     }
-    
-    public List<BigDecimal> getProjectedAssetData(String symbol) {
-        log.info("Generating projected price data for {}", symbol);
-        // TODO Implement getProjectedAssetData()
-        return List.of();
-    }
 
+    public List<BigDecimal> getProjectedAssetData(String symbol, int numberOfMonths) {
```
Comment (csirmazbendeguz) [inline src/main/java/com/ndvr/challenge/service/ChallengeService.java:31]: teszt hianyzik
